### PR TITLE
feat: add optional type annotations

### PR DESCRIPTION
**Pull request by good-idea**
### What does this PR do?

Adds the option to type the `data` from a request's response. Plus a small fix, see below.

Example:

```ts
const CURRENT_USER_QUERY = /*GraphQL*/`
	query UserQuery {
		user {
			id
			name
		}
	}
`

interface UserResponse {
	user: {
		id: string,
		username: string
	}
}

const User = () => {
	const { loading, error, data } = useQuery(CURRENT_USER_QUERY)
	if (!loading || !error) return null
	return <p>Hi, {data.user.name}</p>
}

const TypedUser = () => {
	const { loading, error, data } = useQuery<UserResponse>(CURRENT_USER_QUERY)
	if (!loading || !error) return null
	return <p>Hi, {data.user.name}</p> 
	/*                       👆
	 Property 'name' does not exist on type '{ id: string; username: string; }'
	*/
}
```

image

The types fall back to `any`, so it's backwards-compatible.

(I'm somewhat new to both typescript, this should get a review by someone who knows it better) 

Small fix: `graphQLError` should be `graphQLErrors`. (Right?)

https://github.com/nearform/graphql-hooks/blob/b8852a66f1555defe5f7413979a2f1e2e2e85df6/packages/graphql-hooks/index.d.ts#L121

### Checklist

- [x] I have checked the [contributing document](../blob/master/CONTRIBUTING.md)
- [x] I have added or updated any relevant documentation
- [x] I have added or updated any relevant tests

### CODE REVIEW

File: packages/graphql-hooks/index.d.ts [hunk 1]
```diff
@@ -56,7 +62,7 @@ interface ClientOptions {
   fetch?(url: string, options?: object): Promise<object>
   fetchOptions?: object
   logErrors?: boolean
-  onError?({ result, operation }: { operation: Operation, result: Result }): void
+  onError?({ operation: Operation, result: Result }): void
```
Comment (leonardodino) [inline packages/graphql-hooks/index.d.ts]: this is actually a regression of #175
please revert it.

```suggestion
  onError?({ result, operation }: { operation: Operation, result: Result }): void
```

File: packages/graphql-hooks/index.d.ts [hunk 2]
```diff
@@ -105,23 +111,26 @@ interface UseQueryOptions extends UseClientRequestOptions {
   ssr?: boolean
 }
 
-interface UseClientRequestResult {
+interface UseClientRequestResult<Response> {
```
Comment (bmullan91) [inline packages/graphql-hooks/index.d.ts]: @good-idea Any reason we're using `ResponseData` as the generic name in other places, but `Response` here?

Comment (good-idea) [reply packages/graphql-hooks/index.d.ts]: Good catch, thanks!

File: packages/graphql-hooks/index.d.ts (hunk 1 above)
Comment (good-idea) [reply packages/graphql-hooks/index.d.ts]: Oops! I think I had pasted this in from an out of date version. Prettier changed the formatting so I missed the diff.

File: package.json [hunk 3]
```diff
@@ -9,7 +9,7 @@
     "postinstall": "lerna bootstrap --no-ci",
     "coveralls": "cat ./coverage/lcov.info | ./node_modules/coveralls/bin/coveralls.js",
     "eslint": "eslint '**/*.js'",
-    "prettier": "prettier '**/*.js' '**/*.md' --write",
+    "prettier": "prettier '**/*.js' '**/*.md' '**/*.ts' --write",
```
Comment (bmullan91) [inline package.json]: Nice one @good-idea

**Comment by bmullan91** [comment]
@good-idea This has been released `graphql-hooks@3.5.0` 🎉

**Pull request merged**